### PR TITLE
BUG: read_csv not replacing nans before date conversion

### LINKED ISSUE #11019: read_csv: Inconsistent behaviour with parse_dates if the first row is short
I suspect this is related to #6710.

```
import pandas
from io import StringIO
t = "a,b,c,d\n1970-01-01,2,3,4"
s2 = "a,b,c\n1970-01-01,2,3\n1980-01-01,2,3,4"
s = "a,b,c\n1970-01-01,2,3,4"
pandas.read_csv(StringIO(t), parse_dates=['a'])
pandas.read_csv(StringIO(s2), parse_dates=['a'])
pandas.read_csv(StringIO(s), parse_dates=['a'])
```

Calling `read_csv` with `parse_dates=['a']` does the correct thing for t, raises a CParserError for s2, and raises `TypeError:  'NoneType' object has no attribute '__getitem__'` for s. 

I would expect a CParserError for s as well.

Pandas version: 0.16.2

### LINKED ISSUE #26203: read_excel na_values replacement after parse_dates
```python
### Not working
pd.read_excel(
    "test.xlsx",
    na_values={"Test": ['#', 0]},
    parse_dates=["Test"],
    date_parser=lambda x: pd.to_datetime(x, format="%Y-%m-%d"),
)
### Working
pd.read_csv(
    "test.txt",
    na_values={"Test": ['#', 0]},
    parse_dates=["Test"],
    date_parser=lambda x: pd.to_datetime(x, format="%Y-%m-%d"),
)

```
read_excel behaves different from read_csv in replacing NA values and parsing dates.

Test case:
The test.txt and test.xlsx contain the same data, just one column with header "Test" and 5 entries where 0 and # both represent NA values.
```
Test
2012-10-01
0
2015-05-15
#
2017-09-09
```
The first one crashes while trying to parse a "#" character as date.

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #11019 
- [x] closes #26203
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Comment by phofl** [comment]
Not a clue how this could influence mypy in read_xml

**Comment by phofl** [comment]
Ah it is the init file...

File: pandas/io/parsers/base_parser.py [hunk 1]
```diff
@@ -285,6 +288,11 @@ def _validate_parse_dates_presence(self, columns: list[str]) -> None:
             raise ValueError(
                 f"Missing column provided to 'parse_dates': '{missing_cols}'"
             )
+        # Convert positions to actual column names
+        return [
+            col if (isinstance(col, str) or col in columns) else columns[col]
```
Comment (jreback) [inline pandas/io/parsers/base_parser.py:293]: does this fail on duplicate column names?

File: pandas/io/parsers/base_parser.py [hunk 2]
```diff
@@ -556,6 +564,12 @@ def _convert_to_ndarrays(
             else:
                 col_na_values, col_na_fvalues = set(), set()
 
+            if c in self._parse_date_cols:
```
Comment (jreback) [inline pandas/io/parsers/base_parser.py:567]: comment here on why this is would be helpful

File: pandas/io/parsers/base_parser.py (hunk 1 above)
Comment (phofl) [reply pandas/io/parsers/base_parser.py:293]: No dup cols were already removed through mangle_dup_cols here

File: pandas/io/parsers/base_parser.py (hunk 2 above)
Comment (phofl) [reply pandas/io/parsers/base_parser.py:567]: Done

File: pandas/io/parsers/base_parser.py [hunk 3]
```diff
@@ -237,7 +238,7 @@ def _open_handles(
             errors=kwds.get("encoding_errors", "strict"),
         )
 
-    def _validate_parse_dates_presence(self, columns: list[str]) -> None:
+    def _validate_parse_dates_presence(self, columns: list[str]) -> Iterable:
```
Comment (jreback) [inline pandas/io/parsers/base_parser.py:241]: prob should add a Returns now

Comment (phofl) [reply pandas/io/parsers/base_parser.py:241]: good point, thx

**Review by jreback — APPROVED** [review]
lgtm merge on green

**Pull request merged**